presto-hive/src/main/java/io/prestosql/plugin/hive/ViewReaderUtil.java

```diff
@@ -0,0 +1,229 @@
+            implements ViewReader
+    {
+        private final HiveMetastoreClient mscClient;
+        private final TypeManager typeManager;
+
+        public HiveViewDecoder(HiveMetastoreClient mscClient, TypeManager typemanager)
+        {
+            this.mscClient = mscClient;
+            this.typeManager = requireNonNull(typemanager, "metadata is null");
+        }
+
+        @Override
+        public ConnectorViewDefinition decodeViewData(String viewSql, Table table, CatalogName catalogName)
+        {
+            try {
+                HiveToRelConverter hiveToRelConverter = HiveToRelConverter.create(mscClient);
+                RelNode rel = hiveToRelConverter.convertView(table.getDatabaseName(), table.getTableName());
+                RelToPrestoConverter rel2Presto = new RelToPrestoConverter();
+                String prestoSql = rel2Presto.convert(rel);
+                RelDataType rowType = rel.getRowType();
+                List<ConnectorViewDefinition.ViewColumn> columns = new ArrayList<>(rowType.getFieldCount());
+                for (RelDataTypeField field : rowType.getFieldList()) {
+                    log.debug("Adding column %s, %s, %s, %s", field.getName(),
+                            field.getType().getFullTypeString(),
+                            field.getType().getSqlTypeName(),
+                            field);
+
+                    Type type = typeManager.fromSqlType(getTypeString(field.getType()));
+                    ConnectorViewDefinition.ViewColumn column = new ConnectorViewDefinition.ViewColumn(field.getName(), type.getTypeId());
+                    columns.add(column);
+                }
+                return new ConnectorViewDefinition(prestoSql,
+                        Optional.of(catalogName.toString()),
+                        Optional.of(table.getDatabaseName()),
+                        columns,
+                        Optional.ofNullable(table.getParameters().get(TABLE_COMMENT)),
+                        Optional.empty(),
+                        true);
+            }
+            catch (RuntimeException e) {
+                throw new PrestoException(HIVE_UNKNOWN_ERROR,
+                        format("Error decoding view definition for %s.%s: %s. Please report this to ask_dali@linkedin.com",
+                                table.getDatabaseName(),
+                                table.getTableName(),
+                                e.getMessage()),
+                        e);
+            }
+        }
+
+        // Calcite does not provide correct type strings for non-primitive types.
+        // We add custom code here to make it work. Goal is for calcite/coral to handle this
+        private String getTypeString(RelDataType type)
+        {
+            switch (type.getSqlTypeName()) {
+                case ROW: {
+                    Preconditions.checkState(type.isStruct());
+                    StringBuilder sb = new StringBuilder("row(");
```

Comment:
This can be done with a stream
```java
return type.getFieldList().stream()
        .map(field -> field.getName().toLowerCase(ENGLISH) + " " + getTypeString(field.getType()))
        .collect(joining(",", "row(", ")"));